### PR TITLE
Timeslot Calculation Rework - Part of CASEFLOW-1115

### PR DESCRIPTION
**Pull request by charleyf**
[CASEFLOW-1115](https://vajira.max.gov/browse/CASEFLOW-1115) will be done across multiple PRs:
- Backend changes ([16147](https://github.com/department-of-veterans-affairs/caseflow/pull/16147)), adds db columns, falls back to existing behavior for number_of_slots and begins_at (slot_length_minutes has no existing behavior in db).
- Only show slots when scheduling virtual hearings ([16150](https://github.com/department-of-veterans-affairs/caseflow/pull/16150)).
- UI Fixes PR ([16149](https://github.com/department-of-veterans-affairs/caseflow/pull/16149)), stop using HearingTime to fix various bugs (temporary fix until we replace with custom time modal).
- Slot recalculation PR ([16161](https://github.com/department-of-veterans-affairs/caseflow/pull/16161)), branch is based on UI Fixes PR, accept new db values and calculate slots based on them.

### Description
This PR reworks the slot calculation to be much more flexible. As part of that, all of the slot calculation logic has been moved into it's own utils file and reworks the tests.

### CODE REVIEW

File: client/app/hearings/utils.js [hunk 1]
```diff
@@ -496,53 +496,75 @@ export const formatChangeRequestType = (type) => {
...
+
+/**
+ * Method to calculate an array of available time slots, no filled timeslots or hearings are included
+ * @param {string} numberOfSlots  -- Max number of slots to generate
+ * @param {integer} slotLengthMinutes -- The length of each slot in minutes
+ * @param {string} beginsAt  -- Time of first possible slot in "America/New_York" timezone
+ * @param {string} roTimezone -- Timezone like 'America/Los_Angeles' of the ro
+ * @param {array} hearings    -- List of hearings scheduled for a specific date
+ **/
+const calculateAvailableTimeslots = ({ numberOfSlots, slotLengthMinutes, beginsAt, roTimezone, hearings }) => {
```
Comment (sahalliburton) [inline client/app/hearings/utils.js:558]: Can we re-combine these functions back into 1? It is a bit hard to read what is going on and we should probably only split the functions if there is separate testable functionality that is being provided

**Review by sahalliburton — CHANGES_REQUESTED** [review]
This looks like it is heading in the right direction, however there were a few issues I noticed during testing:

1. I am not sure if this is related to your changes, but I was not able to choose a virtual hearing day for the Oakland RO I had to choose a Video day
2. When selecting a custom time, the options started at 11:15 AM PDT (see screenshot)
3. On the scheduled veterans page the time shown (5:30 PDT) is not the time I chose (8:30 PDT) (see screenshot)
4. On the daily docket page the time is also shown incorrectly as above  (see screenshot)

Let me know if you have any questions or need any help

File: client/app/hearings/utils.js (hunk 1 above)
Comment (charleyf) [reply client/app/hearings/utils.js]: > This looks like it is heading in the right direction, however there were a few issues I noticed during testing:
> 
> 1. I am not sure if this is related to your changes, but I was not able to choose a virtual hearing day for the Oakland RO I had to choose a Video day
> 2. When selecting a custom time, the options started at 11:15 AM PDT (see screenshot)
> 3. On the scheduled veterans page the time shown (5:30 PDT) is not the time I chose (8:30 PDT) (see screenshot)
> 4. On the daily docket page the time is also shown incorrectly as above  (see screenshot)
> 
> Let me know if you have any questions or need any help

Thanks Stephen! I believe these are all addressed.

1. This is not related to this work. There are some restrictions around that I don't quite understand. I confirmed this using the UI.
2. + 4. This was a label formatting problem caused by depending on the Operating System (local) timezone which is why I wasn't seeing it. Reproduced by changing my OS timezone and resolved by changing the label formatting. Thanks for finding this!!!
3. I thought this was the expected behavior, but revisiting it seems clearly wrong. Changed this and it simplified the code nicely as well.

**Comment by charleyf** [comment]
> This looks like it is heading in the right direction, however there were a few issues I noticed during testing:
> 
> 1. I am not sure if this is related to your changes, but I was not able to choose a virtual hearing day for the Oakland RO I had to choose a Video day
> 2. When selecting a custom time, the options started at 11:15 AM PDT (see screenshot)
> 3. On the scheduled veterans page the time shown (5:30 PDT) is not the time I chose (8:30 PDT) (see screenshot)
> 4. On the daily docket page the time is also shown incorrectly as above  (see screenshot)
> 
> Let me know if you have any questions or need any help

Thanks Stephen! I believe these are all addressed.

1. This is not related to this work. There are some restrictions around that I don't quite understand. I confirmed this using the UI.
2. and 4. This was a label formatting problem caused by depending on the Operating System (local) timezone which is why I wasn't seeing it. Reproduced by changing my OS timezone and resolved by changing the label formatting. Thanks for finding this!!!
3. I thought this was the expected behavior, but revisiting it seems clearly wrong. Changed this and it simplified the code nicely as well.

File: client/app/hearings/utils.js (hunk 1 above)
Comment (charleyf) [reply client/app/hearings/utils.js]: I was attempting to simplify things and did not. Thanks for pointing this out, I moved everything back into the loop format they were in.

File: client/app/hearings/utils.js [hunk 2]
```diff
@@ -591,17 +586,29 @@ const combineSlotsAndHearings = ({ roTimezone, availableSlots, scheduledHearings
...
+  const defaultBeginsAt = ro === 'C' ? '09:00' : '08:30';
+  const momentDefaultBeginsAt = moment.tz(defaultBeginsAt, 'HH:mm', 'America/New_York');
+  const defaultSlotLengthMinutes = 60;
 
-  const slotCount = 8;
-  const startTime = ro === 'C' ? '09:00' : '08:30';
   const availableSlots = calculateAvailableTimeslots({
-    slotCount,
-    startTime,
+    numberOfSlots: numberOfSlots ? numberOfSlots : defaultNumberOfSlots,
```
Comment (sahalliburton) [inline client/app/hearings/utils.js:607]: Minor nit, but these could be changed into or assignments like `numberOfSlots: numberOfSlots || defaultNumberOfSlots,`

File: client/app/hearings/utils.js [hunk 3]
```diff
@@ -498,47 +498,42 @@ export const dispositionLabel = (disposition) => HEARING_DISPOSITION_TYPE_TO_LAB
...
     );
-    const slotOutsideTimeRange = slotTime.isBefore(moment.tz(startTime, 'HH:mm', roTimezone));
+    // Make sure that times don't go past midnight into the next day
+    const slotNotOnCorrectDate = !possibleTime.isSame(beginsAt, 'date');
 
-    // Return null if there is a filled time slot, otherwise return the hearingTime
-    if (slotFull || slotOutsideTimeRange) {
-      return null;
-    }
+    // Combine all the conditions that make a slot unavailable
```
Comment (sahalliburton) [inline client/app/hearings/utils.js:528]: Minor nit, but this could be combined into 1 line

**Review by sahalliburton — APPROVED** [review]
LGTM, thanks for making those changes!

**Review by sahalliburton — COMMENTED** [review]
LGTM, but there is still an issue with the custom time dropdown starting at 5:15 local time and only going to 1:45

Screen Shot 2021-04-28 at 8 38 42 AM

**Pull request closed without merging**